File: api/gateway/native/src/routes/v2/bookshelf.ts
import express, { NextFunction, Request, Response } from 'express'
import {
  getBook,
  getBookByIsbn,
  getBookshelf,
  getUser,
  listBookReview,
  listBookshelf,
  listUserWithUserIds,
} from '~/api'
import {
  IGetBookByIsbnInput,
  IGetBookInput,
  IGetBookshelfInput,
  IGetUserInput,
  IListBookReviewInput,
  IListBookshelfInput,
  IListUserByUserIdsInput,
} from '~/types/input'
import {
  IBookOutput,
  IBookshelfListOutput,
  IBookshelfListOutputAuthor,
  IBookshelfListOutputBookshelf,
  IBookshelfOutput,
  IBookshelfOutputAuthor,
  IReviewListOutput,
  IReviewListOutputReview,
  IUserHashOutput,
} from '~/types/output'
import {
  IBookshelfListV2Response,
  IBookshelfListV2ResponseBook,
  IBookshelfListV2ResponseBookshelf,
  IBookshelfV2Response,
  IBookshelfV2ResponseBookshelf,
  IBookshelfV2ResponseReview,
  IBookshelfV2ResponseUser,
} from '~/types/response'
import {
  BookStatus,
  LIST_DEFAULT_LIMIT,
  LIST_DEFAULT_OFFSET,
  LIST_DEFAULT_ORDER_BY,
  LIST_DEFAULT_ORDER_DIRECTION,
} from '~/util'

const router = express.Router()

router.get(
  '/v2/users/:userId/books',
  async (req: Request, res: Response<IBookshelfListV2Response>, next: NextFunction): Promise<void> => {
    const { userId } = req.params
    const { limit, offset } = req.query as { [key: string]: string }

    const userInput: IGetUserInput = {
      id: userId,
    }

    await getUser(req, userInput)
      .then(async () => {
        const bookshelfInput: IListBookshelfInput = {
          userId,
          limit: Number(limit) || LIST_DEFAULT_LIMIT,
          offset: Number(offset) || LIST_DEFAULT_OFFSET,
        }

        return listBookshelf(req, bookshelfInput)
      })
      .then((output: IBookshelfListOutput) => {
        const response: IBookshelfListV2Response = setBookshelfListResponse(output)
        res.status(200).json(response)
      })
      .catch((err: Error) => {
        next(err)
      })
  }
)

router.get(
  '/v2/users/:userId/books/:bookId',
  async (req: Request, res: Response<IBookshelfV2Response>, next: NextFunction): Promise<void> => {
    const { userId, bookId } = req.params
    const { key } = req.query as { [key: string]: string }

    const userInput: IGetUserInput = {
      id: userId,
    }

    await getUser(req, userInput)
      .then(() => {
        switch (key) {
          case 'isbn': {
            const isbn: string = bookId
            return bookByIsbn(req, isbn)
          }
          case 'id':
          default: {
            const id: number = Number(bookId) || 0
            return bookById(req, id)
          }
        }
      })
      .then(async (bookOutput: IBookOutput) => {
        const bookshelfInput: IGetBookshelfInput = {
          userId,
          bookId: bookOutput.id,
        }

        return getBookshelf(req, bookshelfInput)
          .then(async (bookshelfOutput: IBookshelfOutput) => {
            const reviewListInput: IListBookReviewInput = {
              bookId: Number(bookId) || 0,
              limit: LIST_DEFAULT_LIMIT,
              offset: LIST_DEFAULT_OFFSET,
              by: LIST_DEFAULT_ORDER_BY,
              direction: LIST_DEFAULT_ORDER_DIRECTION,
            }

            return listBookReview(req, reviewListInput)
              .then(async (reviewsOutput: IReviewListOutput) => {
                const userIds = reviewsOutput.reviews.map((rv: IReviewListOutputReview): string => {
                  return rv.userId
                })

                const userListInput: IListUserByUserIdsInput = {
                  ids: Array.from(new Set(userIds)),
                }

                return listUserWithUserIds(req, userListInput)
                  .then((usersOutput: IUserHashOutput) => {
                    return setBookshelfResponse(bookOutput, usersOutput, bookshelfOutput, reviewsOutput)
                  })
                  .catch(() => {
                    return setBookshelfResponse(bookOutput, {}, bookshelfOutput, reviewsOutput)
                  })
              })
              .catch((err: Error) => {
                throw err
              })
          })
          .catch(() => {
            return setBookshelfResponse(bookOutput, {})
          })
      })
      .then((response: IBookshelfV2Response) => {
        res.status(200).json(response)
      })
      .catch((err: Error) => {
        next(err)
      })
  }
)

function bookById(req: Request, bookId: number): Promise<IBookOutput> {
  const input: IGetBookInput = {
    bookId,
  }

  return getBook(req, input)
}

function bookByIsbn(req: Request, isbn: string): Promise<IBookOutput> {
  const input: IGetBookByIsbnInput = {
    isbn,
  }

  return getBookByIsbn(req, input)
}

function setBookshelfResponse(
  bookOutput: IBookOutput,
  usersOutput: IUserHashOutput,
  bookshelfOutput?: IBookshelfOutput,
  reviewsOutput?: IReviewListOutput
): IBookshelfV2Response {
  const authorNames = bookOutput.authors.map((item: IBookshelfOutputAuthor): string => {
    return item.name
  })

  const authorNameKanas = bookOutput.authors.map((item: IBookshelfOutputAuthor): string => {
    return item.nameKana
  })

  const reviews = reviewsOutput?.reviews.map(
    (item: IReviewListOutputReview): IBookshelfV2ResponseReview => {
      const user: IBookshelfV2ResponseUser = {
        id: '',
        username: 'unknown',
        thumbnailUrl: '',
      }

      if (usersOutput[item.userId]) {
        const { id, username, thumbnailUrl } = usersOutput[item.userId]

        user.id = id
        user.username = username
        user.thumbnailUrl = thumbnailUrl
      }

      return {
        id: item.id,
        impression: item.impression,
        createdAt: item.createdAt,
        updatedAt: item.updatedAt,
        user,
      }
    }
  )

  const response: IBookshelfV2Response = {
    id: bookOutput.id,
    title: bookOutput.title,
    titleKana: bookOutput.titleKana,
    description: bookOutput.description,
    isbn: bookOutput.isbn,
    publisher: bookOutput.publisher,
    publishedOn: bookOutput.publishedOn,
    thumbnailUrl: bookOutput.thumbnailUrl,
    rakutenUrl: bookOutput.rakutenUrl,
    size: bookOutput.rakutenSize,
    author: authorNames.join('/'),
    authorKana: authorNameKanas.join('/'),
    createdAt: bookOutput.createdAt,
    updatedAt: bookOutput.updatedAt,
    reviews: reviews || [],
    reviewLimit: reviewsOutput?.limit || 0,
    reviewOffset: reviewsOutput?.offset || 0,
    reviewTotal: reviewsOutput?.total || 0,
  }

  if (bookshelfOutput && bookshelfOutput.status !== BookStatus.none) {
    const bookshelf: IBookshelfV2ResponseBookshelf = {
      status: BookStatus[bookshelfOutput.status],
      readOn: bookshelfOutput.readOn,
      reviewId: bookshelfOutput.reviewId === 0 ? bookshelfOutput.reviewId : undefined,
      createdAt: bookshelfOutput.createdAt,
      updatedAt: bookshelfOutput.updatedAt,
    }

    response.bookshelf = bookshelf
  }

  return response
}

function setBookshelfListResponse(output: IBookshelfListOutput): IBookshelfListV2Response {
  const books = output.bookshelves.map(
    (bs: IBookshelfListOutputBookshelf): IBookshelfListV2ResponseBook => {
      const authorNames = bs.book.authors.map((item: IBookshelfListOutputAuthor): string => {
        return item.name
      })

      const authorNameKanas = bs.book.authors.map((item: IBookshelfListOutputAuthor): string => {
        return item.nameKana
      })

      const bookshelf: IBookshelfListV2ResponseBookshelf = {
        status: BookStatus[bs.status],
        readOn: bs.readOn,
        reviewId: bs.reviewId === 0 ? bs.reviewId : undefined,
        createdAt: bs.createdAt,
        updatedAt: bs.updatedAt,
      }

      const book: IBookshelfListV2ResponseBook = {
        id: bs.book.id,
        title: bs.book.title,
        titleKana: bs.book.titleKana,
        description: bs.book.description,
        isbn: bs.book.isbn,
        publisher: bs.book.publisher,
        publishedOn: bs.book.publishedOn,
        thumbnailUrl: bs.book.thumbnailUrl,
        rakutenUrl: bs.book.rakutenUrl,
        size: bs.book.rakutenSize,
        author: authorNames.join('/'),
        authorKana: authorNameKanas.join('/'),
        createdAt: bs.book.createdAt,
        updatedAt: bs.book.updatedAt,
        bookshelf,
      }

      return book
    }
  )

  const response: IBookshelfListV2Response = {
    books,
    limit: output.limit,
    offset: output.offset,
    total: output.total,
  }

  return response
}

export default router

File: api/gateway/native/internal/response/v2/bookshelf.go
package v2

import (
	"strings"

	"github.com/calmato/gran-book/api/gateway/native/internal/entity"
)

// 本棚の書籍情報
type BookshelfResponse struct {
	ID           int64               `json:"id"`           // 書籍ID
	Title        string              `json:"title"`        // タイトル
	TitleKana    string              `json:"titleKana"`    // タイトル(かな)
	Description  string              `json:"description"`  // 説明
	Isbn         string              `json:"isbn"`         // ISBN
	Publisher    string              `json:"publisher"`    // 出版社名
	PublishedOn  string              `json:"publishedOn"`  // 出版日
	ThumbnailURL string              `json:"thumbnailUrl"` // サムネイルURL
	RakutenURL   string              `json:"rakutenUrl"`   // 楽天ショップURL
	Size         string              `json:"size"`         // 楽天書籍サイズ
	Author       string              `json:"author"`       // 著者名一覧
	AuthorKana   string              `json:"author_kana"`  // 著者名一覧(かな)
	Bookshelf    *BookshelfBookshelf `json:"bookshelf"`    // ユーザーの本棚情報
	Reviews      []*BookshelfReview  `json:"reviewsList"`  // レビュー一覧
	ReviewLimit  int64               `json:"reviewLimit"`  // レビュー取得上限
	ReviewOffset int64               `json:"reviewOffset"` // レビュー取得開始位置
	ReviewTotal  int64               `json:"reviewTotal"`  // レビュー検索一致件
	CreatedAt    string              `json:"createdAt"`    // 登録日時
	UpdatedAt    string              `json:"updatedAt"`    // 更新日時
}

func NewBookshelfResponse(
	bs *entity.Bookshelf,
	b *entity.Book,
	rs entity.Reviews,
	um map[string]*entity.User,
	reviewLimit, reviewOffset, reviewTotal int64,
) *BookshelfResponse {
	return &BookshelfResponse{
		ID:           b.Id,
		Title:        b.Title,
		TitleKana:    b.TitleKana,
		Description:  b.Description,
		Isbn:         b.Isbn,
		Publisher:    b.Publisher,
		PublishedOn:  b.PublishedOn,
		ThumbnailURL: b.ThumbnailUrl,
		RakutenURL:   b.RakutenUrl,
		Size:         b.RakutenSize,
		Author:       strings.Join(b.AuthorNames(), "/"),
		AuthorKana:   strings.Join(b.AuthorNameKanas(), "/"),
		CreatedAt:    b.CreatedAt,
		UpdatedAt:    b.UpdatedAt,
		Bookshelf:    newBookshelfBookshelf(bs),
		Reviews:      newBookshelfReviews(rs, um),
		ReviewLimit:  reviewLimit,
		ReviewOffset: reviewOffset,
		ReviewTotal:  reviewTotal,
	}
}

type BookshelfBookshelf struct {
	Status    string `json:"status"`    // 読書ステータス
	ReadOn    string `json:"readOn"`    // 読み終えた日
	ReviewID  int64  `json:"reviewId"`  // レビューID
	CreatedAt string `json:"createdAt"` // 登録日時
	UpdatedAt string `json:"updatedAt"` // 更新日時
}

func newBookshelfBookshelf(bs *entity.Bookshelf) *BookshelfBookshelf {
	return &BookshelfBookshelf{
		Status:    bs.Status().Name(),
		ReadOn:    bs.ReadOn,
		ReviewID:  bs.ReviewId,
		CreatedAt: bs.CreatedAt,
		UpdatedAt: bs.UpdatedAt,
	}
}

type BookshelfReview struct {
	ID         int64          `json:"id"`         // レビューID
	Impression string         `json:"impression"` // 感想
	User       *BookshelfUser `json:"user"`       // 投稿者
	CreatedAt  string         `json:"createdAt"`  // 登録日時
	UpdatedAt  string         `json:"updatedAt"`  // 更新日時
}

func newBookshelfReview(r *entity.Review, u *entity.User) *BookshelfReview {
	return &BookshelfReview{
		ID:         r.Id,
		Impression: r.Impression,
		User:       newBookshelfUser(u),
		CreatedAt:  r.CreatedAt,
		UpdatedAt:  r.UpdatedAt,
	}
}

func newBookshelfReviews(rs entity.Reviews, um map[string]*entity.User) []*BookshelfReview {
	res := make([]*BookshelfReview, len(rs))
	for i, r := range rs {
		u := um[r.UserId]
		res[i] = newBookshelfReview(r, u)
	}
	return res
}

type BookshelfUser struct {
	ID           string `json:"id"`           // ユーザーID
	Username     string `json:"username"`     // 表示名
	ThumbnailURL string `json:"thumbnailUrl"` // サムネイルURL
}

func newBookshelfUser(u *entity.User) *BookshelfUser {
	if u == nil {
		return &BookshelfUser{
			Username: "unknown",
		}
	}

	return &BookshelfUser{
		ID:           u.Id,
		Username:     u.Username,
		ThumbnailURL: u.ThumbnailUrl,
	}
}

// 本棚の書籍一覧
type BookshelfListResponse struct {
	Books  []*BookshelfListBook `json:"booksList"` // 書籍一覧
	Limit  int64                `json:"limit"`     // 取得上限数
	Offset int64                `json:"offset"`    // 取得開始位置
	Total  int64                `json:"total"`     // 検索一致数
}

func NewBookshelfListResponse(
	bss entity.Bookshelves, bm map[int64]*entity.Book, limit, offset, total int64,
) *BookshelfListResponse {
	return &BookshelfListResponse{
		Books:  newBookshelfListBooks(bss, bm),
		Limit:  limit,
		Offset: offset,
		Total:  total,
	}
}

type BookshelfListBook struct {
	ID           int64                   `json:"id"`           // 書籍ID
	Title        string                  `json:"title"`        // タイトル
	TitleKana    string                  `json:"titleKana"`    // タイトル(かな)
	Description  string                  `json:"description"`  // 説明
	Isbn         string                  `json:"isbn"`         // ISBN
	Publisher    string                  `json:"publisher"`    // 出版社名
	PublishedOn  string                  `json:"publishedOn"`  // 出版日
	ThumbnailURL string                  `json:"thumbnailUrl"` // サムネイルURL
	RakutenURL   string                  `json:"rakutenUrl"`   // 楽天ショップURL
	Size         string                  `json:"size"`         // 楽天書籍サイズ
	Author       string                  `json:"author"`       // 著者名一覧
	AuthorKana   string                  `json:"authorKana"`   // 著者名一覧(かな)
	Bookshelf    *BookshelfListBookshelf `json:"bookshelf"`    // ユーザーの本棚情報
	CreatedAt    string                  `json:"createdAt"`    // 登録日時
	UpdatedAt    string                  `json:"updatedAt"`    // 更新日時
}

func newBookshelfListBook(bs *entity.Bookshelf, b *entity.Book) *BookshelfListBook {
	return &BookshelfListBook{
		ID:           b.Id,
		Title:        b.Title,
		TitleKana:    b.TitleKana,
		Description:  b.Description,
		Isbn:         b.Isbn,
		Publisher:    b.Publisher,
		PublishedOn:  b.PublishedOn,
		ThumbnailURL: b.ThumbnailUrl,
		RakutenURL:   b.RakutenUrl,
		Size:         b.RakutenSize,
		Author:       strings.Join(b.AuthorNames(), "/"),
		AuthorKana:   strings.Join(b.AuthorNameKanas(), "/"),
		CreatedAt:    b.CreatedAt,
		UpdatedAt:    b.UpdatedAt,
		Bookshelf:    newBookshelfListBookshelf(bs),
	}
}

func newBookshelfListBooks(bss entity.Bookshelves, bm map[int64]*entity.Book) []*BookshelfListBook {
	res := make([]*BookshelfListBook, 0, len(bss))
	for _, bs := range bss {
		b, ok := bm[bs.BookId]
		if !ok {
			continue
		}

		res = append(res, newBookshelfListBook(bs, b))
	}
	return res
}

type BookshelfListBookshelf struct {
	Status    string `json:"status"`    // 読書ステータス
	ReadOn    string `json:"readOn"`    // 読み終えた日
	ReviewID  int64  `json:"reviewId"`  // レビューID
	CreatedAt string `json:"createdAt"` // 登録日時
	UpdatedAt string `json:"updatedAt"` // 更新日時
}

func newBookshelfListBookshelf(bs *entity.Bookshelf) *BookshelfListBookshelf {
	return &BookshelfListBookshelf{
		Status:    bs.Status().Name(),
		ReadOn:    bs.ReadOn,
		ReviewID:  bs.ReviewId,
		CreatedAt: bs.CreatedAt,
		UpdatedAt: bs.UpdatedAt,
	}
}
